pkg/tekton/metapipeline/client.go

```diff
@@ -0,0 +1,323 @@
+	pipelineName := tekton.PipelineResourceNameFromGitInfo(gitInfo, branchIdentifier, c.Context, tekton.MetaPipeline, tektonClient, ns)
+	buildNumber, err := tekton.GenerateNextBuildNumber(tektonClient, jxClient, ns, gitInfo, branchIdentifier, retryDuration, c.Context)
+	if err != nil {
+		return errors.Wrap(err, "unable to determine next build number")
+	}
+
+	log.Logger().Debugf("creating meta pipeline CRDs for build %s of repository '%s/%s'", buildNumber, gitInfo.Organisation, gitInfo.Name)
+
+	extendingApps, err := GetExtendingApps(jxClient, ns)
+	if err != nil {
+		return err
+	}
+
+	if c.versionsDir == "" {
+		c.versionsDir, err = c.gitCloneVersionStream(jxClient, ns)
+		if err != nil {
+			return err
+		}
+
+		defer os.RemoveAll(c.versionsDir)
+	}
+
+	crdCreationParams := CRDCreationParameters{
+		Namespace:        ns,
+		Context:          c.Context,
+		PipelineName:     pipelineName,
+		ResourceName:     resourceName,
+		PipelineKind:     pipelineKind,
+		BuildNumber:      buildNumber,
+		GitInfo:          *gitInfo,
+		BranchIdentifier: branchIdentifier,
+		PullRef:          *pullRefs,
+		SourceDir:        defaultCheckoutDir,
+		PodTemplates:     podTemplates,
+		ServiceAccount:   c.ServiceAccount,
+		Labels:           c.CustomLabels,
+		EnvVars:          c.CustomEnvs,
+		DefaultImage:     c.DefaultImage,
+		Apps:             extendingApps,
+		VersionsDir:      c.versionsDir,
+	}
+	tektonCRDs, err := CreateMetaPipelineCRDs(crdCreationParams)
+	if err != nil {
+		return errors.Wrap(err, "failed to generate Tekton CRDs for meta pipeline")
+	}
+	// record the results in the struct for the case this command is called programmatically (HF)
+	c.Results = *tektonCRDs
+
+	err = c.handleResult(tektonClient, jxClient, tektonCRDs, buildNumber, branchIdentifier, *pullRefs, ns, gitInfo)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (c *Client) validateArguments() error {
+	// default values
+	if c.DefaultImage == "" {
+		c.DefaultImage = syntax.DefaultContainerImage
+	}
+	if c.ServiceAccount == "" {
+		c.ServiceAccount = "tekton-bot"
+	}
+	if c.OutDir == "" {
+		c.OutDir = "out"
+	}
+
+	if c.SourceURL == "" {
+		return util.MissingOption(sourceURLOptionName)
+	}
+	if c.PullRefs == "" {
+		return util.MissingOption(pullRefOptionName)
+	}
+
+	_, err := util.ExtractKeyValuePairs(c.CustomLabels, "=")
+	if err != nil {
+		return errors.Wrap(err, "unable to parse custom labels")
+	}
+	_, err = util.ExtractKeyValuePairs(c.CustomEnvs, "=")
+	if err != nil {
+		return errors.Wrap(err, "unable to parse custom environment variables")
+	}
+	return nil
+}
+
+func (c *Client) handleResult(tektonClient tektonclient.Interface,
+	jxClient jxclient.Interface,
+	tektonCRDs *tekton.CRDWrapper,
+	buildNumber string,
+	branch string,
+	pullRefs prow.PullRefs,
+	ns string,
+	gitInfo *gits.GitRepository) error {
+
+	pipelineActivity := tekton.GeneratePipelineActivity(buildNumber, branch, gitInfo, &pullRefs, tekton.MetaPipeline)
```

Comment:
This is the part we can improve by having a more fine-grained interface  opposed to a single _Create_.

Reply:
sounds fine - though the immediate need is a single `Create`. We can add more fine grained methods later right?